### ISSUE TITLE
More help with jq/json

### ISSUE THREAD

**Issue by brialparker** [umd-coding-workshop/website#142]
I've been trying to test out what @peichman-umd helped me out with at Coding Workshop yesterday. I'm running into some errors, and I suspect they have to do with the fact that jq is just not fond of this particular flavor of json that comes out of the ArchivesSpace API. 

I've posted the test files I'm working with [here](https://github.com/brialparker/ArchivesSpace_scripts), including the json output and the script. Right now, I'm just trying to get it to echo the uri to see if it works, but am getting errors like 

`
parse error: Invalid numeric literal at line 2, column 0
curl_accession.sh: line 4: uri: command not found
`

I've tried little changes, but am pretty stumped. Is this just the source json I'm working with? does jq not like the <<<?

**Comment by jwestgard** [comment]
I've never really used jq, so I'm not sure, but I suspect the issue is with quoting.  The shell is going to interpret the "uri" as uri, and maybe that's the problem?

For what it's worth, you could do this in Python fairly easily, if the shell wrapper ends up being to fragile.  I know recommending a different language is just about the least helpful sort of advice...

**Comment by peichman-umd** [comment]
The "curl_accession.sh: line 4: uri: command not found" is due to the fact
that when you assign a variable in bash, you cannot put spaces around the
"="; that line of your script needs to be uri=$(jq ."uri" <<<"$line" | sed
's/\"//g')

As for the parse errors, I think that is due to the fact that the for loop
you are using is splitting the input on whitepsace and not newlines, so jq
is getting just JSON fragments. I tried a while read loop instead, and that
worked better:

while read line; do
    # parse out uri, do curl, etc.
done <lit_ids2.json

The command "read" reads a line from STDIN and places it in the variable
you specify (here, "line"). That is why you need the <lit_ids2.json input
redirection at the end of the loop. With this loop, I only get 1 JSON parse
error.

**Comment by brialparker** [comment]
Well then. That works, at least for echoing the lines themselves (except now realizing upon trying to execute the curl that the server whose API I am trying to talk to doesn't have jq. dangit.

**Comment by brialparker** [comment]
Now I'm wondering if I can just output all of the curl statements to a file, line by line, and then just try to loop through that using the API...

**Comment by jwestgard** [comment]
This is the Archivesspace server, right?  And the problem is that the API is only open to localhost, correct?  If so, I think writing the curl commands to a file is a good solution.  In theory, though, it ought to be possible to install jq onto the server without sudo access, but that might be more trouble than it is worth if you can use the method you proposed.

**Comment by brialparker** [comment]
Correct, this is ArchivesSpace, and yes, I think it's only open to localhost. Yeah, I hit a wall from looking at it too much, so I'm gonna step away until tomorrow. But I feel fairly confident I can make something work...

**Comment by peichman-umd** [comment]
That is definitely a legitimate strategy. You've graduated to writing
programs that write programs :-)

For extra fanciness, you can do it as a pipe, and not write out the curl
commands to a file. Assuming that your curl_accession.sh script now writes
out the curl commands to STDOUT, you can do this:

./curl_accession.sh | ssh apiserver bash -s

This pipes the output of curl_accession.sh over ssh to "apiserver" (fill in
your hostname here). The "-s" argument to bash (on the server side) tells
it to read commands from STDIN.

**Comment by jwestgard** [comment]
So, Peter's given you a viable workaround in addition to the perfectly good option your originally proposed of writing the URLs to a file.  Here's a third option that I think could work: use Python 2 with only modules from the standard libraries.  I think this would run on the server without any additional installations:

{code}
#!/usr/bin/env python
import json, urllib2
HOST = "http://aspacestage.lib.umd.edu:8089"
with open('lit_ids2.json', 'r') as f:
    for line in f:
        json_obj = json.loads(line)
        uri = HOST + json_obj['uri']
        r = urllib2.urlopen(uri, data=json_obj)
        print(r.getcode())
{code}

**Comment by brialparker** [comment]
well, I feel that a txt file with lines like:

`
curl -H "X-ArchivesSpace-Session: $TOKEN" -d {"lock_version":0,"suppressed":false,"title":"Vincent Godfrey Burns papers","display_string":"Vincent Godfrey Burns papers","publish":false,"content_description":"Broadside of Burns's "Life Is Far More Than This," inscribed to Irena Ames.","accession_date":"2011-03-30","restrictions_apply":false,"access_restrictions":false,"use_restrictions":false,"created_by":"blparker","last_modified_by":"blparker","create_time":"2015-08-13T16:34:42Z","system_mtime":"2016-04-06T23:08:39Z","user_mtime":"2015-08-13T16:34:42Z","id_0":"2011","id_1":"092","id_2":"LIT","acquisition_type":"Gift","resource_type":"papers","jsonmodel_type":"accession","external_ids":[],"related_accessions":[],"classifications":[],"subjects":[],"linked_events":[],"extents":[{"lock_version":0,"number":"1","created_by":"blparker","last_modified_by":"blparker","create_time":"2015-08-13T16:34:42Z","system_mtime":"2015-08-13T16:34:42Z","user_mtime":"2015-08-13T16:34:42Z","portion":"whole","extent_type":"Items","jsonmodel_type":"extent"}],"dates":[],"external_documents":[],"rights_statements":[],"deaccessions":[],"related_resources":[],"linked_agents":[{"role":"source","terms":[],"ref":"/agents/people/2873"}],"instances":[],"uri":"/repositories/2/accessions/1634","repository":{"ref":"/repositories/2"},"collection_management":{"lock_version":0,"rights_determined":false,"created_by":"blparker","last_modified_by":"blparker","create_time":"2015-08-13T16:34:42Z","system_mtime":"2015-08-13T16:34:42Z","user_mtime":"2015-08-13T16:34:42Z","processing_priority":"high_7","jsonmodel_type":"collection_management","external_ids":[],"uri":"/repositories/2/collection_management/497","repository":{"ref":"/repositories/2"}}} 'http://aspacestage.lib.umd.edu:8089'
`

and a shell script like 
`
while read line; do
    echo `$line`
done <lit_id_update.txt`

should work... But...no....

**Comment by peichman-umd** [comment]
Are you trying to print the lines of the lit_id_update.txt or execute them?

**Comment by brialparker** [comment]
I'm trying to execute, so i'm probably doing it completely wrong. I only now figured out how to escape the stupid backticks in markdown.

**Comment by peichman-umd** [comment]
If you are intending to execute it, remember that the text file of curl
commands you are creating *is* a bash script.

**Comment by brialparker** [comment]
ah, thanks. I'll have to revisit this tomorrow. I feel like I'm _this_ close now....

**Comment by brialparker** [comment]
I'm trying to take a stab at 

./curl_accession.sh | ssh apiserver bash -s

which seems to get me the closest, but I'm wondering if there are some access issues. Generally speaking, it's refusing the connection, and I'm wondering if aspacestage will even let me in via this route....

**Comment by peichman-umd** [comment]
Are you able to ssh in to aspacestage normally? Maybe a username or
password issue? I have ssh keys set up for most of the servers I work with
which makes it smoother to do things like pipe output across ssh (since it
doesn't prompt me for passwords).

**Comment by brialparker** [comment]
That is possible, but there's also a $TOKEN issue that I'm not sure how to handle. Plus, I have to login twice to the server. Once to ssh into it, then another to login to the admin account which allows me API access. I'm wondering if this is not the best option then. 

I just can't figure out how to loop through those curl statements, basically...

**Comment by jwestgard** [comment]
We can try a few things at the coding workshop today if you'll be there for that.

**Comment by jwestgard** [comment]
Argh, one of those days in Coding Workshop.  We ran into the following issues with the various approaches:
1. Shell script writes curl to file; file copied to server and run: we ended up in the seventh circle of shell-quoting Hades, between needing to quote the curl command, but expand certain variables when writing the file, and others when running it, and also needing to maintain the quoting of the JSON.
2. Run Python on server: because we only have the stock Python that comes with RHEL available on the server, had to use urllib2 to handle HTTP, and we couldn't get the session token that Aspace requires to work properly.  In a perfect world, we'd have installed the requests module, but no sudo access on the staging server, let alone production.
3. Run jq on the server -- this is probably the best solution in the long run, but again no permissions to do that now, and so it seemed easier to try 1&2. Now that those ways are both blocked, perhaps we should revisit.
